components/org.wso2.carbon.identity.application.authenticator.fido2.endpoint/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/endpoint/impl/FinishRegistrationApiServiceImpl.java

```diff
@@ -0,0 +1,70 @@
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.application.authenticator.fido2.core.WebAuthnService;
+import org.wso2.carbon.identity.application.authenticator.fido2.endpoint.FinishRegistrationApiService;
+import org.wso2.carbon.identity.application.authenticator.fido2.endpoint.common.Constants;
+import org.wso2.carbon.identity.application.authenticator.fido2.exception.FIDO2AuthenticatorClientException;
+import org.wso2.carbon.identity.application.authenticator.fido2.exception.FIDO2AuthenticatorServerException;
+
+import java.text.MessageFormat;
+import javax.ws.rs.core.Response;
+
+import static org.wso2.carbon.identity.application.authenticator.fido2.endpoint.common.Util.getErrorDTO;
+
+/**
+ * FinishRegistrationApiServiceImpl class is used to complete FIDO2 device registration.
+ */
+public class FinishRegistrationApiServiceImpl extends FinishRegistrationApiService {
+
+    private static final Log LOG = LogFactory.getLog(FinishRegistrationApiServiceImpl.class);
+
+    @Override
+    public Response finishRegistrationPost(String challengeResponse) {
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(MessageFormat.format("Received finish registration  challenge response: {0}",
+                    challengeResponse));
+        }
+        try {
+            WebAuthnService webAuthnService = new WebAuthnService();
+            webAuthnService.finishRegistration(challengeResponse);
+        } catch (FIDO2AuthenticatorClientException e) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Client error while FIDO2 device finish registration.", e);
+            }
+            if (e.getErrorCode().equals(Constants.ErrorMessages
+                    .ERROR_CODE_FINISH_REGISTRATION_USERNAME_AND_CREDENTIAL_ID_EXISTS.getCode())) {
+                return Response.status(Response.Status.CONFLICT).entity(getErrorDTO(Constants.ErrorMessages
+                        .ERROR_CODE_FINISH_REGISTRATION_USERNAME_AND_CREDENTIAL_ID_EXISTS)).build();
+            } else if (e.getErrorCode().equals(Constants.ErrorMessages.ERROR_CODE_FINISH_REGISTRATION_INVALID_REQUEST
+                    .getCode())) {
+                return Response.status(Response.Status.BAD_REQUEST).entity(getErrorDTO(Constants.ErrorMessages
+                        .ERROR_CODE_FINISH_REGISTRATION_INVALID_REQUEST, challengeResponse)).build();
+            }
+        } catch (FIDO2AuthenticatorServerException e) {
+            LOG.error("Unexpected server exception while finishing FIDO2 device registration with challenge " +
+                    "response: " + challengeResponse, e);
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity(getErrorDTO(Constants.ErrorMessages
+                    .ERROR_CODE_FINISH_REGISTRATION)).build();
+        }
+        return Response.ok().entity(challengeResponse).build();
```

Comment:
Move this into try block, update the other similar usages

Reply:
Moved.